LingoClash/LingoClash
# Set up fake data provider

### Description



#### Data Managers 
Data managers handle all db operations ie. CRUD 

#### Data Provider 
Data provider handles all HTTP requests to the server. `DataProvider` protocol is created so that more such providers can be extended in the future. For instance, I have created a `JSONServerDataProvider` to support HTTP requests to json-server's REST API, inherited by `FakeDataProvider` to tailor to LingoClashDevServer (which is inherently a json-server). In the future, FireBaseDataProvider, GraphQLDataProvider, etc, could be added. Inspired by [React-admin's Data Provider](https://marmelab.com/react-admin/DataProviders.html)

- Created a BookDataManager to handle all operations relating to books
- Created a JSONServerDataProvider to handle HTTP requests to LingoClashDevServer, powered by [json-server](https://www.npmjs.com/package/json-server)
- Introduced PromiseKit since I can't live without Promise ><; Please read the [doc](https://github.com/mxcl/PromiseKit/blob/v6/Documentation/GettingStarted.md) to understand how it works.

#### JSONServerDataProvider 

This Data Provider fits REST APIs powered by [JSON Server](https://github.com/typicode/json-server)

| Method             | API calls                                                                                               |
| ------------------ | ------------------------------------------------------------------------------------------------------- |
| `getList`          | `GET http://my.api.url/posts?_sort=title&_order=ASC&_start=0&_end=24&title=bar`                         |
| `getOne`           | `GET http://my.api.url/posts/123`                                                                       |
| `getMany`          | `GET http://my.api.url/posts?id=123&id=456&id=789`     |
| `getManyReference` | `GET http://my.api.url/posts?author_id=345`                                                             |
| `create`           | `POST http://my.api.url/posts`                                                                      |
| `update`           | `PUT http://my.api.url/posts/123`                                                                       |
| `updateMany`       | `PUT http://my.api.url/posts/123`, `PUT http://my.api.url/posts/456`, `PUT http://my.api.url/posts/789` |
| `delete`           | `DELETE http://my.api.url/posts/123`                                                                    |


### Type of change


### How to test



- Run `pod install` to install all dependencies.
- Clone LingoClashDevServer and `cd` into it.
- Run `yarn install` to install all dependencies. 
- Run `yarn run json:server` to start the backend server. 
- Fire up Xcode and Run the simulator.
- Go to Login Page and observe that books are fetched and properly logged into the console. 

### Checklist
